src/decl.cpp

```diff
@@ -141,6 +141,114 @@ static const Type *lApplyTypeQualifiers(int typeQualifiers, const Type *type, So
     return type;
 }
 
+///////////////////////////////////////////////////////////////////////////
+// Attributes
+
+AttrArgument::AttrArgument() : kind(ATTR_ARG_UNKNOWN), intVal(0), stringVal() {}
+AttrArgument::AttrArgument(int64_t i) : kind(ATTR_ARG_UINT32), intVal(i), stringVal() {}
+AttrArgument::AttrArgument(const std::string &s) : kind(ATTR_ARG_STRING), intVal(0), stringVal(s) {}
+
+void AttrArgument::Print() const {
+    switch (kind) {
+    case ATTR_ARG_UINT32:
+        printf("(%" PRId64 ")", intVal);
+        break;
+    case ATTR_ARG_STRING:
+        printf("(\"%s\")", stringVal.c_str());
+        break;
+    case ATTR_ARG_UNKNOWN:
+        printf("(unknown)");
+        break;
+    }
+}
+
+Attribute::Attribute(const std::string &n) : name(n), arg() {}
+Attribute::Attribute(const std::string &n, AttrArgument a) : name(n), arg(a) {}
+Attribute::Attribute(const Attribute &a) : name(a.name), arg(a.arg) {}
+
+bool Attribute::IsKnownAttribute() const {
+    // Known/supported attributes.
+    static std::unordered_map<std::string, int> lKnownParamAttrs = {
+        {"noescape", 1},
+        {"address_space", 1},
+    };
+
+    if (lKnownParamAttrs.find(name) != lKnownParamAttrs.end()) {
+        return true;
+    }
+
+    return false;
+}
+
+void Attribute::Print() const {
+    printf("%s", name.c_str());
+    arg.Print();
+}
+
+AttributeList::AttributeList() {}
+
+AttributeList::AttributeList(const AttributeList &attrList) {
+    auto attrs = attrList.attributes;
+    for (unsigned int i = 0; i < attrs.size(); ++i) {
```

Comment:
Why `unsigned int`?
Use range-based for instead: `for (auto i : attrs.size())`.

If you want to be an old school, use `std::size_t`, which is the type returned by `vector::size()`.

Reply:
No specific reasons. Replaced with range-based syntax all loops.